### PR TITLE
Add dynamic client registration

### PR DESCRIPTION
**Pull request by jasonpaulos**
I've started using this library with [Solid](https://github.com/solid/node-solid-server) to authenticate against [inrupt.net](https://inrupt.net/) and [solid.community](https://solid.community/), and since Solid is decentralized it requires apps to [dynamically register](https://openid.net/specs/openid-connect-registration-1_0.html) with the provider. Both AppAuth-Android and AppAuth-iOS provide support for client registration, so I added the Javascript and native code required to take advantage of that. I would really like to get this accepted so that others don't have to start from scratch with this too!

## Description

I've added a new library function, `register`, which can be used to obtain a client id and secret for subsequent calls to `authorize` and other library functions.

I've tested the functionality on iOS and Android, and everything is working properly.

I did end up modifying the eslint settings, since it was giving me errors when using `== null` to check if parameters had been passed to `register`, and since it was limiting the number of statements that could be used in an arrow function, which affected the tests.

## Steps to verify

```javascript
import { register } from 'react-native-app-auth';

const res = await register({
  issuer: 'https://inrupt.net',
  redirectUrls: ['com.myapp.id:/oauthredirect']
});
// res is now { clientId: '...', clientSecret: '...', ... }
````
Let me know if there's anything I can do to help get this approved.

### CODE REVIEW

File: android/src/main/java/com/rnappauth/RNAppAuthModule.java [hunk 1]
```diff
@@ -130,6 +136,75 @@ public void onFetchConfigurationCompleted(
...
+                        redirectUris,
+                        responseTypes,
+                        grantTypes,
+                        subjectType,
+                        tokenEndpointAuthMethod,
+                        additionalParametersMap,
+                        promise
+                );
+            } catch (Exception e) {
+                promise.reject("Failed to refresh token", e.getMessage());
```
Comment (kadikraman) [inline android/src/main/java/com/rnappauth/RNAppAuthModule.java]:
```suggestion
                promise.reject("registration_failed", e.getMessage());
```

File: android/src/main/java/com/rnappauth/RNAppAuthModule.java [hunk 2]
```diff
@@ -130,6 +136,75 @@ public void onFetchConfigurationCompleted(
...
+        } else {
+            final Uri issuerUri = Uri.parse(issuer);
+            AuthorizationServiceConfiguration.fetchFromUrl(
+                    buildConfigurationUriFromIssuer(issuerUri),
+                    new AuthorizationServiceConfiguration.RetrieveConfigurationCallback() {
+                        public void onFetchConfigurationCompleted(
+                                @Nullable AuthorizationServiceConfiguration fetchedConfiguration,
+                                @Nullable AuthorizationException ex) {
+                            if (ex != null) {
+                                promise.reject("Failed to fetch configuration", getErrorMessage(ex));
```
Comment (kadikraman) [inline android/src/main/java/com/rnappauth/RNAppAuthModule.java]:
```suggestion
                                promise.reject("service_configuration_fetch_error", getErrorMessage(ex));
```

File: android/src/main/java/com/rnappauth/RNAppAuthModule.java [hunk 3]
```diff
@@ -345,6 +416,64 @@ public void onTokenRequestCompleted(
...
+        RegistrationRequest registrationRequest = registrationRequestBuilder.build();
+
+        AuthorizationService.RegistrationResponseCallback registrationResponseCallback = new AuthorizationService.RegistrationResponseCallback() {
+            @Override
+            public void onRegistrationRequestCompleted(@Nullable RegistrationResponse response, @Nullable AuthorizationException ex) {
+                if (response != null) {
+                    WritableMap map = RegistrationResponseFactory.registrationResponseToMap(response);
+                    promise.resolve(map);
+                } else {
+                    promise.reject("Failed to refresh token", getErrorMessage(ex));
```
Comment (kadikraman) [inline android/src/main/java/com/rnappauth/RNAppAuthModule.java]:
```suggestion
                    promise.reject("token_refresh_failed", getErrorMessage(ex));
```

File: ios/RNAppAuth.m [hunk 4]
```diff
@@ -33,6 +33,50 @@ - (dispatch_queue_t)methodQueue
...
+                            subjectType: subjectType
+                tokenEndpointAuthMethod: tokenEndpointAuthMethod
+                   additionalParameters: additionalParameters
+                                resolve: resolve
+                                 reject: reject];
+    } else {
+        [OIDAuthorizationService discoverServiceConfigurationForIssuer:[NSURL URLWithString:issuer]
+                                                            completion:^(OIDServiceConfiguration *_Nullable configuration, NSError *_Nullable error) {
+                                                                if (!configuration) {
+                                                                    reject(@"RNAppAuth Error", [error localizedDescription], error);
```
Comment (kadikraman) [inline ios/RNAppAuth.m]:
```suggestion
                                                                    reject(@"service_configuration_fetch_error", [error localizedDescription], error);
```

File: ios/RNAppAuth.m [hunk 5]
```diff
@@ -163,6 +207,45 @@ + (nullable NSString *)codeChallengeS256ForVerifier:(NSString *)codeVerifier {
...
+                                  tokenEndpointAuthMethod:tokenEndpointAuthMethod
+                                     additionalParameters:additionalParameters];
+    
+    [OIDAuthorizationService performRegistrationRequest:request
+                                             completion:^(OIDRegistrationResponse *_Nullable response,
+                                                          NSError *_Nullable error) {
+                                                 if (response) {
+                                                     resolve([self formatRegistrationResponse:response]);
+                                                 } else {
+                                                     reject(@"RNAppAuth Error", [error localizedDescription], error);
```
Comment (kadikraman) [inline ios/RNAppAuth.m]:
```suggestion
                                                     reject(@"registration_failed", [error localizedDescription], error);
```

File: android/src/main/java/com/rnappauth/RNAppAuthModule.java (hunk 3 above)
Comment (jasonpaulos) [reply android/src/main/java/com/rnappauth/RNAppAuthModule.java]: I think this should be `registration_failed` as well

**Comment by jasonpaulos** [comment]
Thank you @kadikraman! I have updated the error messages as you suggested.

**Comment by kadikraman** [comment]
Released in [v5.0.0](https://github.com/FormidableLabs/react-native-app-auth/releases/tag/v5.0.0) 🎉

**Pull request merged**